### PR TITLE
ENH: Add DataFrame tabular repr

### LINKED ISSUE #1604: DataFrame tabular repr
Like normal `pd.DataFrame`, it's nice to check `DataFrame` overview via tabular format. 

My idea is adding a table with `divisions`, `columns` and `dtype` to existing string repr as below (without calling `.compute()`). It should also work on console accordingly.

Any ideas?

### PR DESCRIPTION
**Pull request by sinhrks**
closes #1604. Fixed to be distinguishable from pandas output.

### CODE REVIEW

**Comment by mrocklin** [comment]
@jcrist any comments on this?

**Comment by mrocklin** [comment]
Taking another look at this (sorry for the long silence).  Some feedback based on the following example:

image

Some subjective thoughts:

*  I think we should show fewer empty lines for divisions.  Maybe four or five
*  I think that we should remove the text-based representation above the tabular repr.  
*  I think that we should include the number of partitions somewhere, perhaps where it currently says "divisions"
*  It would be nice to also state how many tasks are in the computation.  I'm not sure exactly where this would go, but we could consider using some of the empty space below the dtypes.
*  If the computation of `df.head` is cheap we could consider adding the first few lines as a sample

    ```python
    head = df.head(compute=False)
    head.dask = head._optimize(head.dask, head._keys())
    if len(head.dask) < 10:
        ...
    ```

    I'm not sure about this though.  This sort of guess work can get us in trouble.  Just throwing it out there as a thought.

**Comment by sinhrks** [comment]
Sorry not to follow this up.  updated based on your suggestions.

- [x] show fewer empty lines for divisions. Maybe four or five
- [x] remove the text-based representation above the tabular repr.
- [x] include the number of partitions somewhere



I think it's nice to display dask key name and number of tasks. Do we already have a function to count current tasks?

**Comment by mrocklin** [comment]
> Do we already have a function to count current tasks?

I would just use `len(self.dask)`

**Comment by mrocklin** [comment]
Trying this out locally now.  It feels very nice to me.

**Comment by mrocklin** [comment]
Oh, and I see that it works nicely as a text repr as well.

**Comment by sinhrks** [comment]
> I would just use len(self.dask)

Ah you're right. I've tried to count something other than dask tasks. 

Fixed to include it to the html / string repr. Note that ``to_string`` only contains data repr, but I do not have specific preference.

File: dask/dataframe/core.py [hunk 1]
```diff
@@ -300,6 +290,33 @@ def _elemwise(self):
         return elemwise
 
     @property
+    def _repr_data(self):
+        raise NotImplementedError
+
+    @property
+    def _repr_name(self):
+        return self._name if len(self._name) < 10 else self._name[:7]
```
Comment (mrocklin) [inline dask/dataframe/core.py]: If we wanted to use more than 7 or 10 characters then we might want to use the `dask.utils.key_split` function introduced in https://github.com/dask/dask/pull/1919 .

**Comment by mrocklin** [comment]
I've added the use of key_split to determine the name.  Any objections @sinhrks ?  

I would like to merge this soon.

**Comment by sinhrks** [comment]
thx, no objections of course:)

**Comment by mrocklin** [comment]
Merged.  Thanks @sinhrks !  I think this change will make several people happy.

**Pull request merged**